### ISSUE TITLE
bug in interpretation of dtMinToleratedTimeScale in adapt time step using dtAdaptControlProbe

### ISSUE THREAD

**Issue by garkenyon** [PetaVision/OpenPV#42]
the line in HyPerCol for testing the maximum allowed error under the control of a dtAdaptControlProbe:

```
    if (timeScaleProbe > 0 && timeScaleProbe < dtMinToleratedTimeScale) {
```

is not formulated correctly.  I suspect when the adaptive time step was refactored to use the new dtAdaptControlProbe the meaning of timeScaleProbe got redefined but the above check wasn't correctly updated.

timeScaleProbe is the current value of the objective cost function, whereas in the old adapt time step method each error layer was returning 1/error^2 I believe.

anyway, the condition now should be testing whether timeScaleProbe exceeds a maximum value (1/dtMinToleratedTimeScale), not being below a minimum value.  in the current code, a perfect reconstruction would throw and error every time!  in fact, that's what's happening in my current AWS run, which is how I discovered this bug.

**Comment by garkenyon** [comment]
a quick workaround if this is an issue for someone is to set dtMinToleratedTimeScale = 0.0, which effectively turns off all checking

**Comment by peteschultz** [comment]
The way that dtAdaptControlProbe is used in the system tests that use it (DtAdaptControllerTest, LCATest, MomentumLCATest, and TotalEnergyTest), timeScaleProbe is not the cost function, but the quotient (energy of input image)/(cost function).  So a reconstruction with a cost function of zero would have timeScaleProbe of infinity and would never throw the error.  Should the time scale be (cost function)/(energy of input image) in these tests instead?

(In checking this, I found that in LCATest and MomentumLCATest, the InputEnergyColProbe isn't connected to anything, so the HyPerCol_timescales.txt file is all zeros.  This seems like a mistake)

**Comment by peteschultz** [comment]
Adding an L2NormProbe for the input energy in LCATest, as I mentioned above, causes that test to fail when restarting from checkpoint, so there seems to be a checkpointing bug regarding adaptive timesteps.

**Comment by garkenyon** [comment]
It looks like one big difference here is that I'm not using a QuotientColProbe in my current LCA runs with adaptive time steps.  I'm just using a ColumnEnergyProbe.  to compute the adaptation in the time step, I'm using the percentage decrease in timeScaleProbe  returned by dtAdaptControlProbe.  see HyPerCol::adaptTimeScaleExp1stOrder.  

I think that if we're using a dtAdaptControlProbe, which should be required (the old method is deprecated I believe), then indeed I think dtAdaptControlProbe should simply return the energy and the system tests should use the normalized energy (divided by the Image L2 norm)

sorry about the confusion.  It's possible I messed up something in the checkpointing of the adaptive time step dynamic variables.  to my knowledge, no system test has been successfully implemented to check the new adaptive time step, which is what I guess you're trying to rectify now.

**Comment by peteschultz** [comment]
I fixed the checkpointing bug I mentioned above, and changed the system tests to use adaptTimeScaleExp1stOrder: https://github.com/PetaVision/OpenPV/commit/b34e1a72928259b39ffbc7fc395d5deba93a8d3a.

Regarding the original issue, setting dtAdaptControlFlag to true and setting dtAdaptController to NULL is indeed deprecated.  Are we also deprecating dtAdaptTimeScale() in favor of dtAdaptTimeScaleExp1stOrder()?

**Comment by peteschultz** [comment]
Closed as fixed by commit 611897d009cfabed4d045d4f689824ca6edfa869.
